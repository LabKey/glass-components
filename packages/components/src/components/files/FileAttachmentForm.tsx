/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { Button } from 'react-bootstrap';
import { List, Map } from 'immutable';

import { FormSection } from '../base/FormSection';
import { Progress } from '../base/Progress';
import { FileAttachmentContainer } from './FileAttachmentContainer';
import { FileGridPreviewProps, FilePreviewGrid } from './FilePreviewGrid';
import { LoadingSpinner } from '../base/LoadingSpinner';
import { convertRowDataIntoPreviewData, fileMatchesAcceptedFormat, fileSizeLimitCompare } from './actions';
import { InferDomainResponse } from '../base/models/model';
import { inferDomainFromFile } from '../base/actions';
import { FileSizeLimitProps } from './models';



interface FileAttachmentFormProps {
    acceptedFormats?: string // comma-separated list of allowed extensions i.e. '.png, .jpg, .jpeg'
    // map between extension and SizeLimitProps.  Use "all" as the key for limits that apply to all formats.
    // "all" limits will be overridden by limits for a specific extension.
    sizeLimits?: Map<string, FileSizeLimitProps>
    sizeLimitsHelpText?: React.ReactNode
    showAcceptedFormats?: boolean
    allowDirectories?: boolean
    allowMultiple?: boolean
    cancelText?: string
    initialFileNames?: Array<string>
    initialFiles?: {[key:string]: File}
    index?: number
    label?: string
    labelLong?: string
    onCancel?: () => any
    onFileChange?: (files: Map<string, File>) => any
    onFileRemoval?: (attachmentName: string) => any
    onSubmit?: (files: Map<string, File>) => any
    isSubmitting?: boolean
    showButtons?: boolean
    showLabel?: boolean
    showProgressBar?: boolean
    submitText?: string
    previewGridProps?: FileGridPreviewProps
    templateUrl?: string
    compact?: boolean
}

interface State {
    attachedFiles: Map<string, File>
    errorMessage?: string
    previewData?: List<Map<string, any>>
    previewStatus?: string
}

export class FileAttachmentForm extends React.Component<FileAttachmentFormProps, State> {

    static defaultProps = {
        acceptedFormats: '',
        showAcceptedFormats: true,
        allowDirectories: true,
        allowMultiple: true,
        cancelText: 'Cancel',
        label: 'Attachments',
        labelLong: 'Select file or drag and drop here.',
        onCancel: undefined,
        onSubmit: undefined,
        showButtons: false,
        showLabel: true,
        showProgressBar: false,
        submitText: 'Upload',
        compact: false
    };

    constructor(props?: FileAttachmentFormProps) {
        super(props);

        if (props.allowMultiple && props.previewGridProps) {
            console.warn('Showing the file preview grid is only supported for single file upload.')
        }

        this.state = {
            attachedFiles: Map<string, File>(),
            errorMessage: undefined,
            previewStatus: undefined
        };
    }

    componentWillMount() {
        this.initPreviewData(this.props)
    }

    componentWillReceiveProps(nextProps: FileAttachmentFormProps) {
        if (this.props.previewGridProps !== nextProps.previewGridProps) {
            this.initPreviewData(nextProps);
        }
    }

    initPreviewData(props: FileAttachmentFormProps) {
        let previewData;
        if (props.previewGridProps && props.previewGridProps.initialData) {
            previewData = convertRowDataIntoPreviewData(props.previewGridProps.initialData.get('data'), props.previewGridProps.previewCount);
            this.setState(() => ({previewData}));
        }
    }

    determineFileSize(): number {
        const { attachedFiles } = this.state;

        return attachedFiles.reduce((total, file) =>( total += file.size), 0);
    }

    handleFileChange = (fileList: {[key: string]: File}) => {
        const { onFileChange, sizeLimits } = this.props;
        const attachedFiles = this.state.attachedFiles.merge(fileList);

        this.setState(() => ({attachedFiles}), () => {

            if (this.isShowPreviewGrid()) {
                // if showing preview, there can be only one attached file
                const sizeCheck = fileSizeLimitCompare(attachedFiles.valueSeq().first(), sizeLimits);
                if (!sizeCheck.isOversizedForPreview)
                    this.uploadDataFileForPreview();
                else {
                    this.setState(() => ({
                        errorMessage: "This file is too large to be previewed. The maximum size allowed for previewing files of this type is " + sizeCheck.limits.maxPreviewSize.displayValue
                    }))
                }
            }

            if (onFileChange) {
                onFileChange(attachedFiles)
            }
        });
    };

    handleFileRemoval = (attachmentName: string) => {
        const { onFileRemoval } = this.props;

        this.setState(() => ({
            attachedFiles: this.state.attachedFiles.remove(attachmentName),
            previewData: undefined,
            previewStatus: undefined,
            errorMessage: undefined
        }), () => {
            if (onFileRemoval) {
                onFileRemoval(attachmentName);
            }
        });
    };

    handleSubmit = () => {
        const { onSubmit } = this.props;

        if (onSubmit)
            onSubmit(this.state.attachedFiles);
        // clear out attached files once they have been submitted.
        this.setState(() => ({
            attachedFiles: Map<string, File>()
        }));
    };

    renderButtons() {
        const { cancelText, onCancel, submitText } = this.props;

        return (
            <div className="row top-spacing bottom-spacing">
                <div className="col-xs-6">
                    <Button
                        onClick={onCancel}
                        bsStyle="default"
                        title={cancelText}>
                        {cancelText}
                    </Button>
                </div>
                <div className="col-xs-6">
                    <div className="pull-right">
                        <Button
                            className={"file-form-submit-btn"}
                            onClick={this.handleSubmit}
                            bsStyle="success"
                            disabled={this.state.attachedFiles.size == 0}
                            title={submitText}>
                            {submitText}
                        </Button>
                    </div>
                </div>
            </div>
        )
    }

    isShowPreviewGrid() {
        return !this.props.allowMultiple && this.props.previewGridProps;
    }

    shouldShowPreviewGrid() {
        const { errorMessage, previewData, previewStatus } = this.state;
        return errorMessage || previewStatus || previewData;
    }

    renderPreviewGrid() {
        const { previewGridProps } = this.props;
        const { errorMessage, previewData, previewStatus } = this.state;

        if (!this.shouldShowPreviewGrid()) {
            return;
        }

        if (previewData || errorMessage) {
            return (
                <FilePreviewGrid
                    {...previewGridProps}
                    data={previewData}
                    errorMsg={errorMessage}
                />
            )
        }
        else if (previewStatus) {
            return (
                <div className={"margin-top"}>
                    <LoadingSpinner msg={previewStatus}/>
                </div>
            )
        }
    }

    updatePreviewStatus(previewStatus: string) {
        this.setState(() => ({previewStatus}));
    }

    updateErrors(errorMessage: string) {
        this.setState(() => ({errorMessage}));
    }

    uploadDataFileForPreview() {
        const { previewGridProps } = this.props;
        const { attachedFiles } = this.state;

        //just take the first file, since we only support 1 file at this time
        const file = attachedFiles.first();

        // check if this usage has a set of formats which are supported for preview
        if (previewGridProps.acceptedFormats) {
            const fileCheck = fileMatchesAcceptedFormat(file.name, previewGridProps.acceptedFormats);
            // if the file extension doesn't match the accepted preview formats, return without trying to get preview data
            if (!fileCheck.get('isMatch')) {
                return;
            }
        }

        this.updatePreviewStatus("Uploading file...");

        inferDomainFromFile(file, previewGridProps.previewCount)
            .then((response: InferDomainResponse) => {
                this.updatePreviewStatus(null);

                if (!previewGridProps.skipPreviewGrid) {
                    if (response.data.size > 1) {
                        const previewData = convertRowDataIntoPreviewData(response.data, previewGridProps.previewCount, response.fields);
                        this.setState(() => ({previewData}));
                        this.updateErrors(null);
                    }
                    else {
                        this.updateErrors('No data found in the attached file.');
                    }
                }

                if (previewGridProps.onPreviewLoad) {
                    previewGridProps.onPreviewLoad(response);
                }
            })
            .catch((reason) => {
                this.updateErrors(reason);
            })
    }

    shouldRenderAcceptedFormats(): boolean {
        const { acceptedFormats, showAcceptedFormats } = this.props;
        return acceptedFormats && showAcceptedFormats && !this.shouldShowPreviewGrid();
    }

    renderAcceptedFormats() {
        return (
            <div className={"file-form-formats"}>
                <strong>Supported formats include: </strong>{this.props.acceptedFormats}
            </div>
        )
    }

    shouldRenderTemplateButton(): boolean {
        const { templateUrl } = this.props;
        return templateUrl && templateUrl.length > 0 && !this.shouldShowPreviewGrid();
    }

    renderTemplateButton() {

        return (
            <Button
                bsStyle={'info'}
                title={'Download Template'}
                href={this.props.templateUrl}
            >
                <span className="fa fa-download"/> Template
            </Button>
        )
    }

    renderFooter() {
        if (!this.shouldRenderAcceptedFormats() && !this.shouldRenderTemplateButton()) {
            return;
        }

        return (
            <div className="row">
                <div className="col-md-9">
                    {this.shouldRenderAcceptedFormats() && this.renderAcceptedFormats()}
                </div>
                <div className="col-md-3">
                    <div className={'pull-right'}>
                        {this.shouldRenderTemplateButton() && this.renderTemplateButton()}
                    </div>
                </div>
            </div>
        )
    }

    render() {
        const {
            acceptedFormats,
            allowDirectories,
            allowMultiple,
            initialFileNames,
            initialFiles,
            label,
            labelLong,
            showButtons,
            showLabel,
            showProgressBar,
            sizeLimits,
            sizeLimitsHelpText,
            isSubmitting,
            compact
        } = this.props;

        return (
            <>
                <span className="translator--toggle__wizard">
                    <FormSection
                        iconSpacer={false}
                        label={label}
                        showLabel={showLabel}
                    >
                        <FileAttachmentContainer
                            index={this.props.index}
                            acceptedFormats={acceptedFormats}
                            allowDirectories={allowDirectories}
                            handleChange={this.handleFileChange}
                            handleRemoval={this.handleFileRemoval}
                            initialFileNames={initialFileNames}
                            initialFiles={initialFiles}
                            allowMultiple={allowMultiple}
                            sizeLimits={sizeLimits}
                            sizeLimitsHelpText={sizeLimitsHelpText}
                            labelLong={labelLong}
                            compact={compact}
                        />
                    </FormSection>
                </span>
                {this.renderPreviewGrid()}
                {showProgressBar && (
                    <Progress
                        estimate={this.determineFileSize() * .1}
                        modal={true}
                        title="Uploading"
                        toggle={isSubmitting}/>
                )}
                {this.renderFooter()}
                {showButtons && this.renderButtons()}
            </>
        )
    }
}
