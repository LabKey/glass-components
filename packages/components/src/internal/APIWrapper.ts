import { SamplesAPIWrapper, SamplesServerAPIWrapper, getSamplesTestAPIWrapper } from './components/samples/APIWrapper';
import {
    PicklistAPIWrapper,
    PicklistServerAPIWrapper,
    getPicklistTestAPIWrapper,
} from './components/picklist/APIWrapper';
import {
    getSecurityTestAPIWrapper,
    SecurityAPIWrapper,
    ServerSecurityAPIWrapper,
} from './components/security/APIWrapper';
import {
    DomainPropertiesAPIWrapper,
    getDomainPropertiesTestAPIWrapper,
} from './components/domainproperties/APIWrapper';

export interface ComponentsAPIWrapper {
    picklist: PicklistAPIWrapper;
    samples: SamplesAPIWrapper;
    security: SecurityAPIWrapper;
    domain: DomainPropertiesAPIWrapper;
}

export function getDefaultAPIWrapper(): ComponentsAPIWrapper {
    return {
        picklist: new PicklistServerAPIWrapper(),
        samples: new SamplesServerAPIWrapper(),
        security: new ServerSecurityAPIWrapper(),
        domain: new DomainPropertiesAPIWrapper(),
    };
}

/**
 * Note: Intentionally does not use jest.fn() to avoid jest becoming an implicit external package dependency.
 */
export function getTestAPIWrapper(
    mockFn = (): any => () => {},
    overrides: Partial<ComponentsAPIWrapper> = {}
): ComponentsAPIWrapper {
    return {
        picklist: getPicklistTestAPIWrapper(mockFn, overrides.picklist),
        samples: getSamplesTestAPIWrapper(mockFn, overrides.samples),
        security: getSecurityTestAPIWrapper(mockFn, overrides.security),
        domain: getDomainPropertiesTestAPIWrapper(mockFn, overrides.domain),
        ...overrides,
    };
}
