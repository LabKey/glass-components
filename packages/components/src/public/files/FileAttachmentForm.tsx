/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { PureComponent, ReactNode } from 'react';
import { List, Map } from 'immutable';
import classNames from 'classnames';

import { FileAttachmentContainer } from '../../internal/components/files/FileAttachmentContainer';
import {
    convertRowDataIntoPreviewData,
    fileMatchesAcceptedFormat,
    fileSizeLimitCompare,
    getFileExtension,
} from '../../internal/components/files/actions';
import { FilePreviewGrid } from '../../internal/components/files/FilePreviewGrid';
import { SimpleResponse } from '../../internal/components/files/models';

import { LoadingSpinner } from '../../internal/components/base/LoadingSpinner';
import { FormSection } from '../../internal/components/base/FormSection';
import { Progress } from '../../internal/components/base/Progress';

import { inferDomainFromFile } from '../InferDomainResponse';

import { TemplateDownloadButton } from './TemplateDownloadButton';
import { FileSizeLimitProps, FileGridPreviewProps } from './models';

export interface FileAttachmentFormProps {
    acceptedFormats?: string; // comma-separated list of allowed extensions i.e. '.png, .jpg, .jpeg'
    allowDirectories?: boolean;
    allowMultiple?: boolean;
    cancelText?: string;
    compact?: boolean;
    fileCountSuffix?: string;
    // map between file extension and the callback function to use instead of the standard uploadDataFileForPreview
    fileSpecificCallback?: Map<string, (file: File) => Promise<SimpleResponse>>;
    includeDirectoryFiles?: boolean;
    index?: number;
    initialFileNames?: string[];
    initialFiles?: Record<string, File>;
    isSubmitting?: boolean;
    label?: ReactNode;
    labelLong?: string;
    onCancel?: () => void;
    onError?: (error: string) => void;
    onFileChange?: (files: Map<string, File>) => void;
    onFileRemoval?: (attachmentName: string, updatedFiles?: Map<string, File>) => void;
    onSubmit?: (files: Map<string, File>) => void;
    previewGridProps?: FileGridPreviewProps;
    ref?: any;
    showAcceptedFormats?: boolean;
    showButtons?: boolean;
    showLabel?: boolean;
    showProgressBar?: boolean;
    // map between extension and SizeLimitProps.  Use "all" as the key for limits that apply to all formats.
    // "all" limits will be overridden by limits for a specific extension.
    sizeLimits?: Map<string, FileSizeLimitProps>;
    sizeLimitsHelpText?: ReactNode;
    submitText?: string;
    templateUrl?: string;
}

interface State {
    attachedFiles: Map<string, File>;
    errorMessage?: string;
    previewData?: List<Map<string, any>>;
    previewStatus?: string;
}

export class FileAttachmentForm extends PureComponent<FileAttachmentFormProps, State> {
    static defaultProps = {
        acceptedFormats: '',
        showAcceptedFormats: true,
        allowDirectories: true,
        includeDirectoryFiles: false,
        allowMultiple: true,
        cancelText: 'Cancel',
        label: 'Attachments',
        labelLong: 'Select file or drag and drop here.',
        onCancel: undefined,
        onSubmit: undefined,
        showButtons: false,
        showLabel: true,
        showProgressBar: false,
        submitText: 'Upload',
        compact: false,
        fileSpecificCallback: undefined,
    };

    fileAttachmentContainerRef: React.RefObject<FileAttachmentContainer>;

    constructor(props?: FileAttachmentFormProps) {
        super(props);

        this.fileAttachmentContainerRef = React.createRef();

        if (props.allowMultiple && props.previewGridProps) {
            console.warn('Showing the file preview grid is only supported for single file upload.');
        }

        this.state = {
            attachedFiles: Map<string, File>(),
            errorMessage: undefined,
            previewStatus: undefined,
        };
    }

    componentDidMount(): void {
        this.initPreviewData();
    }

    componentDidUpdate(prevProps: FileAttachmentFormProps): void {
        if (prevProps.previewGridProps !== this.props.previewGridProps) {
            this.initPreviewData();
        }
    }

    initPreviewData(): void {
        const { previewGridProps } = this.props;

        if (previewGridProps?.initialData) {
            this.setState({
                previewData: convertRowDataIntoPreviewData(
                    previewGridProps.initialData.get('data'),
                    previewGridProps.previewCount
                ),
            });
        }
    }

    determineFileSize = (): number => {
        return this.state.attachedFiles.reduce((total, file) => (total += file.size), 0);
    };

    reportFileOversized = (sizeStr: string): void => {
        const errorMessage = `This file is too large to be previewed. The maximum size allowed for previewing files of this type is ${sizeStr}.`;
        this.setState({ errorMessage });
    };

    handleFileChange = (fileList: Record<string, File>): void => {
        this.setState(
            state => ({ attachedFiles: state.attachedFiles.merge(fileList) }),
            () => {
                const { onFileChange, sizeLimits, fileSpecificCallback, allowMultiple } = this.props;
                const { attachedFiles } = this.state;

                if (!allowMultiple) {
                    // currently only supporting 1 file for processing contents
                    const firstFile = attachedFiles.valueSeq().first();
                    const sizeCheck = fileSizeLimitCompare(firstFile, sizeLimits);

                    const fileTypeFn = fileSpecificCallback?.get(getFileExtension(firstFile.name));
                    if (fileTypeFn) {
                        if (!sizeCheck.isOversized) {
                            fileTypeFn(firstFile)
                                .then(res => {
                                    this.updateErrors(res.success ? null : res.msg);
                                })
                                .catch(res => {
                                    this.updateErrors(res.msg);
                                });
                        } else {
                            this.reportFileOversized(sizeCheck.limits.maxSize.displayValue);
                        }
                    } else if (this.isShowPreviewGrid()) {
                        if (!sizeCheck.isOversizedForPreview) {
                            this.uploadDataFileForPreview();
                        } else {
                            this.reportFileOversized(sizeCheck.limits.maxPreviewSize.displayValue);
                        }
                    }
                }

                onFileChange?.(attachedFiles);
            }
        );
    };

    handleFileRemoval = (attachmentName: string): void => {
        this.setState(
            state => ({
                attachedFiles: state.attachedFiles.remove(attachmentName),
                errorMessage: undefined,
                previewData: undefined,
                previewStatus: undefined,
            }),
            () => {
                this.props.onFileRemoval?.(attachmentName, this.state.attachedFiles);
            }
        );
    };

    handleSubmit = (): void => {
        this.props.onSubmit?.(this.state.attachedFiles);

        // clear out attached files once they have been submitted.
        this.setState({ attachedFiles: Map<string, File>() });
    };

    renderButtons(): ReactNode {
        const { cancelText, onCancel, submitText, compact } = this.props;

        const button = (
            <button
                className={classNames('file-form-submit-btn btn btn-success', {
                    'file-form-submit-btn--compact': compact,
                })}
                disabled={this.state.attachedFiles.size === 0}
                onClick={this.handleSubmit}
                title={submitText}
                type="button"
            >
                {submitText}
            </button>
        );

        if (compact) {
            return button;
        }

        return (
            <div className="row top-spacing bottom-spacing">
                <div className="col-xs-6">
                    <button className="btn btn-default" onClick={onCancel} title={cancelText} type="button">
                        {cancelText}
                    </button>
                </div>
                <div className="col-xs-6">
                    <div className="pull-right">{button}</div>
                </div>
            </div>
        );
    }

    isShowPreviewGrid = (): boolean => {
        return !this.props.allowMultiple && !!this.props.previewGridProps;
    };

    shouldShowPreviewGrid = (): boolean => {
        const { errorMessage, previewData, previewStatus } = this.state;
        return !!(errorMessage || previewStatus || previewData);
    };

    renderPreviewGrid(): ReactNode {
        const { previewGridProps } = this.props;
        const { errorMessage, previewData, previewStatus } = this.state;

        if (this.shouldShowPreviewGrid()) {
            if (previewData || errorMessage) {
                return <FilePreviewGrid {...previewGridProps} data={previewData} errorMsg={errorMessage} />;
            } else if (previewStatus) {
                return (
                    <div className="margin-top">
                        <LoadingSpinner msg={previewStatus} />
                    </div>
                );
            }
        }

        return null;
    }

    updatePreviewStatus(previewStatus: string): void {
        this.setState({ previewStatus });
    }

    updateErrors(errorMessage: string): void {
        if (errorMessage) {
            this.props.onError?.(errorMessage);
        }
        this.setState({ errorMessage });
    }

    uploadDataFileForPreview(): void {
        const { previewGridProps } = this.props;
        const { attachedFiles } = this.state;

        // just take the first file, since we only support 1 file at this time
        const file = attachedFiles.first();

        // check if this usage has a set of formats which are supported for preview
        if (previewGridProps.acceptedFormats) {
            const fileCheck = fileMatchesAcceptedFormat(file.name, previewGridProps.acceptedFormats);
            // if the file extension doesn't match the accepted preview formats, return without trying to get preview data
            if (!fileCheck.isMatch) {
                return;
            }
        }

        this.updatePreviewStatus('Uploading file...');

        inferDomainFromFile(file, previewGridProps.previewCount, previewGridProps.domainKindName)
            .then(response => {
                this.updatePreviewStatus(null);

                if (!previewGridProps.skipPreviewGrid) {
                    if (response.data.size > 1) {
                        const previewData = convertRowDataIntoPreviewData(
                            response.data,
                            previewGridProps.previewCount,
                            response.fields
                        );
                        this.setState(() => ({ previewData }));
                        this.updateErrors(null);
                    } else {
                        this.updateErrors('No data found in the attached file.');
                    }
                }

                previewGridProps.onPreviewLoad?.(response, file);
            })
            .catch(reason => {
                this.updateErrors(
                    reason
                        ? reason?.exception || reason
                        : 'There was a problem determining the fields in the uploaded file.  Please check the format of the file.'
                );
            });
    }

    shouldRenderAcceptedFormats(): boolean {
        return !!(this.props.acceptedFormats && this.props.showAcceptedFormats && !this.shouldShowPreviewGrid());
    }

    shouldRenderTemplateButton = (): boolean => {
        return this.props.templateUrl?.length > 0 && !this.shouldShowPreviewGrid();
    };

    render(): React.ReactNode {
        const {
            acceptedFormats,
            allowDirectories,
            includeDirectoryFiles,
            fileCountSuffix,
            allowMultiple,
            index,
            initialFileNames,
            initialFiles,
            label,
            labelLong,
            showButtons,
            showLabel,
            showProgressBar,
            sizeLimits,
            sizeLimitsHelpText,
            isSubmitting,
            compact,
        } = this.props;
        const renderAcceptFormats = this.shouldRenderAcceptedFormats();
        const renderTemplateButton = this.shouldRenderTemplateButton();

        return (
            <>
                <span className="translator--toggle__wizard">
                    <FormSection iconSpacer={false} label={label} showLabel={showLabel}>
                        <div className={classNames({ 'file-upload--one-row': compact })}>
                            <FileAttachmentContainer
                                ref={this.fileAttachmentContainerRef}
                                index={index}
                                acceptedFormats={acceptedFormats}
                                allowDirectories={allowDirectories}
                                includeDirectoryFiles={includeDirectoryFiles}
                                fileCountSuffix={fileCountSuffix}
                                handleChange={this.handleFileChange}
                                handleRemoval={this.handleFileRemoval}
                                initialFileNames={initialFileNames}
                                initialFiles={initialFiles}
                                allowMultiple={allowMultiple}
                                sizeLimits={sizeLimits}
                                sizeLimitsHelpText={sizeLimitsHelpText}
                                labelLong={labelLong}
                                compact={compact}
                            />
                            {compact && showButtons && this.renderButtons()}
                        </div>
                    </FormSection>
                </span>
                {this.renderPreviewGrid()}
                {showProgressBar && (
                    <Progress estimate={this.determineFileSize() * 0.1} modal title="Uploading" toggle={isSubmitting} />
                )}
                {(renderAcceptFormats || renderTemplateButton) && (
                    <div className="row">
                        <div className="col-md-9">
                            {renderAcceptFormats && (
                                <div className="file-form-formats">
                                    <strong>Supported formats include: </strong>
                                    {acceptedFormats}
                                </div>
                            )}
                        </div>
                        <div className="col-md-3">
                            <div className="pull-right">
                                {renderTemplateButton && (
                                    <TemplateDownloadButton templateUrl={this.props.templateUrl} />
                                )}
                            </div>
                        </div>
                    </div>
                )}
                {!compact && showButtons && this.renderButtons()}
            </>
        );
    }
}
