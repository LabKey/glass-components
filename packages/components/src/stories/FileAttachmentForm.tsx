/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
 */
import React from 'react';
import { fromJS, Map } from 'immutable';
import { storiesOf } from '@storybook/react';
import { boolean, number, text, withKnobs } from '@storybook/addon-knobs';
import mock, { proxy } from 'xhr-mock';

import { FileAttachmentForm } from '../components/files/FileAttachmentForm';
import inferDomainJson from '../test/data/property-inferDomainWithSpecimenId.json';
import filePreviewJson from '../test/data/property-getFilePreview.json';

import './stories.scss';

mock.setup();
mock.post(/.*\/property\/inferDomain.*/, {
    status: 200,
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(inferDomainJson)
});
mock.use(proxy);

storiesOf('FileAttachmentForm', module)
    .addDecorator(withKnobs)
    .add('with knobs', () => {
        return (
            <FileAttachmentForm
                showLabel={boolean('Show label?', true)}
                label={text('Label', 'Attachments')}
                labelLong={text('Long label (inside of file attachment drop zone)', 'Select file or drag and drop here')}
                acceptedFormats={text('Accepted formats', '.tsv,.txt,.csv,.xls,.xlsx')}
                showAcceptedFormats={boolean('Show accepted formats?', true)}
                allowDirectories={boolean('Allow directories?', true)}
                allowMultiple={boolean('Allow multiple?', true)}
                showButtons={boolean('Show buttons?', true)}
                cancelText={text('Cancel button text', 'Cancel')}
                submitText={text('Submit button text', 'Upload')}
                templateUrl={text('Download Template URL', '#downloadtemplate')}
                previewGridProps={{
                    previewCount: number('Preview Grid Row Count', 3),
                    acceptedFormats: text('Preview Grid Accepted Formats', '.tsv,.txt,.csv,.xls,.xlsx')
                }}
            />
        )
    })
    .add("with initial file names and preview data", () => {
        return (
            <FileAttachmentForm
                showLabel={true}
                label={'Attachments'}
                labelLong={'Select file or drag and drop here'}
                acceptedFormats={'.tsv,.txt,.csv,.xls,.xlsx'}
                showAcceptedFormats={true}
                allowMultiple={true}
                showButtons={false}
                previewGridProps={{
                    previewCount: number('Preview Grid Row Count', 3),
                    acceptedFormats: text('Preview Grid Accepted Formats', '.tsv,.txt,.csv,.xls,.xlsx'),
                    initialData: fromJS(filePreviewJson)
                }}
                initialFileNames={["test.txt", "other.csv"]}
            />
        )
    })
    .add("with initial files and preview data", () => {
        let files = {};
        files["test1.txt"] = undefined;
        files["test2.txt"] = undefined;
        return (
            <FileAttachmentForm
                showLabel={true}
                label={'Attachments'}
                labelLong={'Select file or drag and drop here'}
                acceptedFormats={'.tsv,.txt,.csv,.xls,.xlsx'}
                showAcceptedFormats={true}
                allowMultiple={true}
                showButtons={false}
                initialFiles={files}
            />
        )
    })
;
