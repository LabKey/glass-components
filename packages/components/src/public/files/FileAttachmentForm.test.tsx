import React from 'react';

import { renderWithAppContext } from '../../internal/test/reactTestLibraryHelpers';
import { TEST_USER_EDITOR } from '../../internal/userFixtures';

import { FileAttachmentForm } from './FileAttachmentForm';

describe('<FileAttachmentForm/>', () => {
    test('no props', () => {
        renderWithAppContext(<FileAttachmentForm />, {
            serverContext: { user: TEST_USER_EDITOR },
        });
        expect(document.querySelector('.fa-download')).not.toBeInTheDocument();
        expect(document.querySelector('.file-formats')).not.toBeInTheDocument();
        expect(document.querySelector('.control-label').textContent).toEqual('Attachments');
        expect(document.querySelector('.file-upload__input').getAttribute('multiple')).toEqual('');
    });

    test('with attributes', () => {
        renderWithAppContext(
            <FileAttachmentForm
                acceptedFormats=".tsv, .xls, .xlsx"
                allowDirectories={false}
                allowMultiple={false}
                label="file attachment"
                templateUrl="#downloadtemplateurl"
            />,
            {
                serverContext: { user: TEST_USER_EDITOR },
            }
        );
        expect(document.querySelector('.fa-download')).toBeInTheDocument();
        expect(document.querySelector('.file-form-formats')).toBeInTheDocument();
        expect(document.querySelector('.control-label').textContent).toEqual('file attachment');
        expect(document.querySelector('.file-upload__input').getAttribute('multiple')).toEqual(null);
    });
});
