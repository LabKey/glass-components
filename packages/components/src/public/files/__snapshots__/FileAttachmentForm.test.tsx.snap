// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileAttachmentForm/> no props 1`] = `
<div>
  <span
    class="translator--toggle__wizard"
  >
    <div
      class="row"
    >
      <div
        class="col-sm-12"
      >
        <label
          class="control-label text-left"
        >
          <strong>
            Attachments
          </strong>
        </label>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-12"
      >
        <div
          class="wizard-row--container"
        >
          <div
            class=""
          >
            <div
              class="file-upload--container block"
            >
              <label
                class="file-upload--label"
                for="fileUpload"
              >
                <i
                  aria-hidden="true"
                  class="fa fa-cloud-upload fa-2x cloud-logo"
                />
                Select file or drag and drop here.
              </label>
              <input
                accept=""
                class="file-upload--input"
                id="fileUpload"
                multiple=""
                name="fileUpload"
                type="file"
              />
            </div>
            <div
              class="file-upload--file-entry-listing"
            />
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;

exports[`<FileAttachmentForm/> with attributes 1`] = `
<div>
  <span
    class="translator--toggle__wizard"
  >
    <div
      class="row"
    >
      <div
        class="col-sm-12"
      >
        <label
          class="control-label text-left"
        >
          <strong>
            file attachment
          </strong>
        </label>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-12"
      >
        <div
          class="wizard-row--container"
        >
          <div
            class=""
          >
            <div
              class="file-upload--container block"
            >
              <label
                class="file-upload--label"
                for="fileUpload"
              >
                <i
                  aria-hidden="true"
                  class="fa fa-cloud-upload fa-2x cloud-logo"
                />
                Select file or drag and drop here.
              </label>
              <input
                accept=".tsv, .xls, .xlsx"
                class="file-upload--input"
                id="fileUpload"
                name="fileUpload"
                type="file"
              />
            </div>
            <div
              class="file-upload--file-entry-listing"
            />
          </div>
        </div>
      </div>
    </div>
  </span>
  <div
    class="row"
  >
    <div
      class="col-md-9"
    >
      <div
        class="file-form-formats"
      >
        <strong>
          Supported formats include: 
        </strong>
        .tsv, .xls, .xlsx
      </div>
    </div>
    <div
      class="col-md-3"
    >
      <div
        class="pull-right"
      >
        <a
          class="btn btn-info undefined"
          href="#downloadtemplateurl"
          rel="noopener noreferrer"
          target="_blank"
          title="Download Template"
        >
          <span
            class="fa fa-download"
          />
           
          Template
        </a>
      </div>
    </div>
  </div>
</div>
`;
